pkg/s3select/input.go

```diff
@@ -0,0 +1,359 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package s3select
+
+import (
+	"bytes"
+	"compress/bzip2"
+	"encoding/csv"
+	"encoding/xml"
+	"io"
+	"strconv"
+	"strings"
+	"time"
+
+	"net/http"
+
+	gzip "github.com/klauspost/pgzip"
+)
+
+var curBuf *bytes.Buffer
+var progressTicker *time.Ticker
+var continuationTimer *time.Timer
+
+// Progress represents a struct that represents the format for XML of the
+// progress messages
+type progress struct {
+	BytesScanned   int64 `xml:"BytesScanned"`
+	BytesProcessed int64 `xml:"BytesProcessed"`
+	BytesReturned  int64 `xml:"BytesReturned"`
+}
+
+// Progress represents a struct that represents the format for XML of the stat
+// messages
+type stats struct {
+	BytesScanned   int64 `xml:"BytesScanned"`
+	BytesProcessed int64 `xml:"BytesProcessed"`
+	BytesReturned  int64 `xml:"BytesReturned"`
+}
+
+// StatInfo is a struct that represents the
+type statInfo struct {
+	BytesScanned   int64
+	BytesReturned  int64
+	BytesProcessed int64
+}
+
+// Input represents a record producing input from a  formatted file or pipe.
+type Input struct {
+	options         *Options
+	reader          *csv.Reader
+	firstRow        []string
+	header          []string
+	minOutputLength int
+	stats           *statInfo
+}
+
+// Options options are passed to the underlying encoding/csv reader.
+type Options struct {
+	// HasHeader when true, will treat the first row as a header row.
+	HasHeader bool
+
+	// FieldDelimiter is the string that fields are delimited by.
+	FieldDelimiter string
+
+	// Comments is the string the first character of a line of
+	// text matches the comment character.
+	Comments string
+
+	// Name of the table that is used for querying
+	Name string
+
+	// ReadFrom is where the data will be read from.
+	ReadFrom io.Reader
+
+	// If true then we need to add gzip reader.
+	// to extract the csv.
+	CompressedGzip bool
+
+	// If true then we need to add bzip reader.
+	// to extract the csv.
+	CompressedBzip bool
+
+	// SQL expression meant to be evaluated.
+	Expression string
+
+	// What the outputted CSV will be delimited by .
+	OutputFieldDelimiter string
+
+	// Size of incoming object
+	StreamSize int64
+
+	// Whether Header is "USE" or another
+	HeaderOpt bool
+}
+
+// NewInput sets up a new Input, the first row is read when this is run.
+// If there is a problem with reading the first row, the error is returned.
+// Otherwise, the returned reader can be reliably consumed with ReadRecord()
+// until ReadRecord() returns nil.
+func NewInput(opts *Options) (*Input, error) {
+	myReader := opts.ReadFrom
+	var tempBytesScanned int64
+	tempBytesScanned = 0
+	curBuf = bytes.NewBuffer(make([]byte, 1000000))
+	if opts.CompressedGzip {
+		tempBytesScanned = opts.StreamSize
+		var err error
+		if myReader, err = gzip.NewReader(opts.ReadFrom); err != nil {
+			return nil, ErrTruncatedInput
+		}
+		if opts.CompressedBzip {
```

Comment:
Either one can be set, you won't have a gzipped bzipped reader, can you check if AWS support this? .. Also, this code flow only sets bzip2 if Gzip is set which is not the case.